src/Locator/RootsFileOrDirectoryLocator.php

```diff
@@ -33,66 +33,89 @@
 
 declare(strict_types=1);
 
-namespace Infection\Finder;
+namespace Infection\Locator;
 
-use Infection\Finder\Exception\LocatorException;
+use function current;
+use const DIRECTORY_SEPARATOR;
 use function Safe\realpath;
 use Symfony\Component\Filesystem\Filesystem;
+use Webmozart\Assert\Assert;
+use Webmozart\PathUtil\Path;
 
 /**
  * @internal
  */
-final class Locator implements LocatorInterface
+final class RootsFileOrDirectoryLocator implements Locator
 {
-    /**
-     * @var string[]
-     */
-    private $paths;
+    private $roots;
+    private $filesystem;
 
     /**
-     * @var Filesystem
+     * @param  string[] $roots
      */
-    private $filesystem;
-
-    public function __construct(array $paths, Filesystem $filesystem)
+    public function __construct(array $roots, Filesystem $filesystem)
     {
-        $this->paths = $paths;
+        Assert::allString($roots);
+
+        $this->roots = $roots;
         $this->filesystem = $filesystem;
     }
 
-    public function locate(string $name): string
+    /**
+     * {@inheritdoc}
+     */
+    public function locate(string $fileFileName): string
     {
-        if ($this->filesystem->isAbsolutePath($name)) {
-            if ($this->filesystem->exists($name)) {
-                return realpath($name);
+        $fileFileName = Path::canonicalize($fileFileName);
```

Comment:
1. What does `Path::canonicalize()` fix? Why do we need it?

2. If it's really needed, could we please not modify original parameter but created a local var instead? (common good practise)

Reply:
1. It is mostly to take care of the trailing `/`. It would be trivial to not depend on `Path` here on linux machines, but I'm less confident about handling it properly on Windows (I don't know if there is any special case I'm not aware of to handle).

2. Noted